### ISSUE TITLE
Talk about Gladiatus performance

### ISSUE THREAD

**Issue by FrutyX** [DinoDevs/GladiatusCrazyAddon#271]
Heya,

I have a problem for a quite some time, this isn't about the servers, I have noticed too that my browser console is giving me sometimes an timeout when working with in-game items (the game does not accept them, page refresh required), and I know that Thanos has already mentioned it on the Glad. forums, but this isn't about it.

My problem is linked to Google Chrome, it's been like this for probably more than 2 months, now, I know you are not like Chrome developers nor Gladiatus developers, I just want to ask, you already helped me with the market crash (ye, that was a GCA issue, but still...), so maybe you'll know something.

Well I don't know how long I have this, but only on Google Chrome - Gladiatus is horribly slow, the page itself loads very slowly, it takes about 4 seconds to load every page, meanwhile while I turn on Microsoft Edge, Gladiatus is insanely fast, the page is loaded immediately, no delay.

Of course, I have tried some basics like clearing cookies, turning off all extensions, reinstalling Chrome, even my Windows was completely deleted and installed again (well, not because of this :D ). The browser console is not giving me any problems.

Maybe this is linked to the server problems, but why Edge is working normally? They are both based on Chromium, almost like clones and yet, Chrome is just very slow. I believe Gladiatus is the only page that does that.

If you know something, anything (possible workaround), just let me know, it's very annoying, I am so used to Chrome and I have everything synced, so I don't want to just swap browsers, if you understand me.

Thank you...

**Comment by GramThanos** [comment]
First, disclaimer, I haven't seen the code of Gladiatus but based on my experience I can see some things that may cause similar problems. So now lets dive into the problem.

I suggest you to open the developer tools (F12) and go to the network tab, then reload your page.
This will let you see the requests to the server and the timing.
Check the list and see that is flagged as "pending" which means is not answered by the server.

Here is an example:
![image](https://user-images.githubusercontent.com/14858959/123445054-3136d600-d5e0-11eb-8c98-a8c7fbba055f.png)
Under the waterfall column you will see what stalls your page.

**Comment by GramThanos** [comment]
An other thing that you can check is whether you have multiple windows open. There is a bug which generates requests to the servers that they haven't fixed for years, if you have 2 tabs opened to the same account and this bug kicks, this may block all the other requests to the server.

**Comment by FrutyX** [comment]
Hmmm, I don't see anything unusual.

But maybe I should have explained myself better.

The page is loaded, it's not like I see a blank page - Gladiatus is loaded and I see all it's contents, but I cannot interact with it, there is a delay before it will let me work with Gladiatus ( 2+- secs), I can see it also with my mouse, my mouse isn't changing to a pointer, only after it is truly "loaded" (same goes with selected menu items, like current page).

According to the Network tab, Gladiatus is loaded, and yet, I wasn't been able to work with it, it's like it is frozen for a second, and then it works normally.

In the menu, you can see how it works, a small GIF is most likely the best way, to show it:

![test](https://user-images.githubusercontent.com/23278552/123452693-f5e7d780-d5de-11eb-8c86-3b5a6bbdfbf3.gif)

A small "freeze", would be the best description.

**Comment by FrutyX** [comment]
After some digging, I found a page, where this problem isn't present - in combat reports (mod=reports&submod=showCombatReport&t=0&reportId=....)

There is a strange behaviour tho, Chrome keeps nonstop requesting a specific image from css, meanwhile Firefox nor Edge are not requesting it, may be useful:

![energie](https://user-images.githubusercontent.com/23278552/123456614-52e58c80-d5e3-11eb-9db0-8ef4be5a1d52.PNG)

**Comment by FrutyX** [comment]
Hold up!
 
I think... it's the menu that is doing this!

When I use the... uhhh, how to put it in english? Tabs? Ah, screenshot:

![tabs](https://user-images.githubusercontent.com/23278552/123458333-39454480-d5e5-11eb-874f-f15c69571a56.PNG)

The webpage is working normally!!! Without a freeze!

When I use the main menu, the same page gets frozen for a while!

**Comment by FrutyX** [comment]
Okay, now, it's your turn.

Do you have any idea why when the page is loaded via the tab menu and not the main menu, it works normally?

**Comment by GramThanos** [comment]
Ok. I understood your problem.
The image problem you spoted was one of the bugs I was talking about.

One thing that I think you should check is if your browser is using hardware acceleration. Check your options and maybe also the chrome://gpu page.

**Comment by FrutyX** [comment]
Hardware acce. is on, I've tried to turn it off and restart my browser, looks like nothing changed (I think I have tried this in the past already).

There are some problems detected (on chrome://gpu), I'll leave a screenshot here for you, not sure if it will help.

![gpu](https://user-images.githubusercontent.com/23278552/123471692-76fe9900-d5f6-11eb-9994-191df11ebd01.PNG)

And about this bug with those requests, I am not having Gladiatus on multiple opened tabs, always only one.

This must be seriously creating some heat on the game servers, I really don't understand, that they don't listen to you.

**Comment by FrutyX** [comment]
So, I actually found on the internet some problems related to chrome, hardware acc. and AMD GPU drivers, unfortunately none of these "fixes" helped.

So this, is maybe completely different and not that much related to the game, I don't know, I can run for example Twitch, Youtube, Facebook and all the sites without problems... but Gladiatus... and since my whole Windows was reinstalled (meaning even my GPU drivers were updated), then I really don't know, probably some combination of Chrome and AMD drivers, maybe.

If you have any ideas, just throw it at me.

**Comment by GramThanos** [comment]
I guess you can compare the same page on edge and on chrome.

I think there is also another tab on the developers tools called profiler that can show you what hangs in terms of javascript and other CPU/GPU related actions.

**Comment by FrutyX** [comment]
The Performance tab in browser tools is very useful. So clearly the longest task usually takes 2s ( that fits), it is under "System", but the activity itself is "Image Decode".

I gotta google for this.

**Comment by FrutyX** [comment]
Well, I found crap.

Why I keep having such dumb normally non-existing problems? This is giving me the market crash flashbacks.

Most of the "fixes" I have seen are just editing Chrome flags, none of them worked, I even updated my GPU driver that was released not long ago.

This is seriously annoying, but I don't want to bother you with it, because clearly this is just a GCA github forum :-)

Anyways, I appreciate the help, thank you.

I'll let you know if I make any progress, but for now, I am dry.


_.... and it is worth to mention, that I have been using ver. 4.3.5 since you created the Market crash fix (on chrome and edge), and I can say, that this version works without any problems (except the bugs that are already reported) :-)._

**Comment by GramThanos** [comment]
You can also try to remove some animations:

You can run this on the dev tools console and test if that page is smoother (you need to run it every time you change page)
```
document.getElementById('banner_event').remove();
document.getElementsByClassName('avatar_costume_animation')[0].remove();
var style = document.createElement('style');
style.innerHTML = '*, *:before, *:after {-o-transition-property: none !important;-moz-transition-property: none !important;-ms-transition-property: none !important;-webkit-transition-property: none !important;transition-property: none !important;-o-transform: none !important;-moz-transform: none !important;-ms-transform: none !important;-webkit-transform: none !important;transform: none !important;-webkit-animation: none !important;-moz-animation: none !important;-o-animation: none !important;-ms-animation: none !important;animation: none !important;}';
document.head.appendChild(style);
```

**Comment by FrutyX** [comment]
Now this is interesting.

I mean, your code did not help, i's still freezing, what is interesting is your idea, a custom script. I could run it for example win Tampermonkey on Gladiatus, but first, I would have to find a code that would have fix that, and for that, I need you, since I am not really experienced in javascript.

But I don't know how to help you, we still know, that the main menu is most likely causing it, could you probably... edit it? Not sure what options we have, but remove from it any interactions like selected menu item, mouse on hover, and this stuff. 

Basically make it just a clickable button without anything else? Because like I've said, when I use the tabs, the page is loaded normally.

It may be worth the shot.

**Comment by FrutyX** [comment]
I went to https://gladiatus.gamerz-bg.com/ which is using a similiar menu, and it seems, that sometimes, it's freezing a bit too, according to performance tool, it's taking 1+-  sec, again image decode.

I also found this https://support.innogames.com/kb/ForgeOfEmpires/en_US/977 , so there is most likely something wrong with Chrome, unfortunately no solution.

**Comment by GramThanos** [comment]
You can also check the https://developer.mozilla.org/en-US/docs/Web/Demos_of_open_web_technologies demos. Maybe then we can pin point the cause of the problem.

**Comment by FrutyX** [comment]
So, I am not making this easy for you, i know.

I looked at most of these and all of them are working smoothly, I even found some browser benchmark tests and found no issues.

I see it, and I can't do anything about it. I guess I will have to wait till it's gonna fix itself (chrome release), well I already waited for  v91 and I got disappointed.

And yes, the long tasks seems to be the freeze.

![task1](https://user-images.githubusercontent.com/23278552/123517165-c64cd400-d69f-11eb-86fc-a6b37c550bca.PNG)
![task2](https://user-images.githubusercontent.com/23278552/123517164-c64cd400-d69f-11eb-85d7-0d681e987a60.PNG)
![task3](https://user-images.githubusercontent.com/23278552/123517163-c5b43d80-d69f-11eb-9566-c7f49d9bce24.PNG)

**Comment by GramThanos** [comment]
I am pretty sure that you found the problem, but we have to find out how to fix it and which exact part of the page is causing it.

Based on this https://stackoverflow.com/questions/15920775/image-decode-times-from-chrome-timeline-dev-tool
I think that it has to be related by something that is animated. I think to pinpoint the problem maybe it is easier if you find a gladiatus page that doesn't have this problem and compare the visual elements with this page. I tried to reproduce the problem but I was not able.

**Comment by FrutyX** [comment]
Hmmm.

When I keep refreshing the same page over and over again, it seems it's okay.

So at least for me, it seems, that this freeze is occuring when the menuitem is changing to an active one (class=menuitem active), when it stays on the same one, the freeze does not occur, so far this is how it behaves.

**Comment by FrutyX** [comment]
Okay, that was a theory.

According to some of my "testing", it seems removing the active and hover element does not help.

I got no clue, unfortunately.

**Comment by GramThanos** [comment]
Try this:

```
var style = document.createElement('style');
style.innerHTML = '.menu-sprite,.menuitem,.menuitem.active,a.menuitem:hover,.menuitem.premium,.menuitem.premium.active,.menuitem.premium:not(.inactive):hover,.submenu .menuitem,.submenu .menuitem.active,.submenu .menuitem:not(.inactive):hover,#submenuhead1,#submenuhead2 {background-position: 0px 0px !important;}';
document.head.appendChild(style);
```

It is ugly, but if it works we may have a fix.

**Comment by FrutyX** [comment]
The performance tool was right.

Image decode... I've blocked all images on Gladiatus, and the problem is gone. I was wrong about the menu.

I gotta find more details, because clearly I can't block all images of course, so far everything was pointing on **/images/bg_main_inner.png** (for you: https://s34-cz.gladiatus.gameforge.com/game/9391/img/ui/layout/bg_main_inner.png), but it is too soon to say for sure.

I need to find a tool/extension, that can block a specific image, or maybe if you can do that with a script, I can just add it's url.
More testing is needed.

**Comment by FrutyX** [comment]
Yea, so I got an idea... I found the problem. I know you mentioned it here somewhere on Github before, I am talking about custom scaling options in Windows.

I am using a custom scaling settings for Windows and clearly Chrome's "image decoder" is most likely resizing those pictures, so they fit, which means it takes a little bit. I currently have it on 120%, when I set it on 100%, it works normally.

And before you throw fire at me, I have custom scaling settings for years and I have been playing Gladiatus normally, this problem just randomly popped up so I had no idea what is happening, probably a bug with Chrome, must be, since other browsers are okay.

Now I am going to find if I can make it work even with my own scaling. Such a crap, I didn't found a word about such problem with scaling on the whole internet. So you better remember it, and if you see someone with similair problems (and not only on Gladiatus), you will know.

**Comment by GramThanos** [comment]
LoL cool! I am glad you found the problem!
Maybe we can find the appropriate webpage and report that to google chrome or chromium.

**Comment by FrutyX** [comment]
I'll see what I can do.

I would like to thank you again for taking your time with this (trying to reproduce, making those codes, advices...) and if you need anything in the future with the addon (like testing or something), just let me know :-).

By the way, are there any options when it comes to editing Gladiatus? I currently don't see much a solution, I mean, I found a way to change Chrome's scaling to default, so Gladiatus is not affected, but it sucks, since I am used to my scaling and it's not very comfortable for me, and something like "scaling only for a specific site" doesn't probably even exists, anyways, I created a thread on the Chrome forums and we'll see what they say.

Note: Simply editing page zoom doesn't do anything.

**Comment by GramThanos** [comment]
No problem mate, any time.

You may use the browser's zoom feature (ctrl+mouse wheel or ctrl + (+)/(-). But it may not be ideal.

**Comment by FrutyX** [comment]
It's time that I'll be on my way, I spammed here long enough.

Found the problem, so I am closing it here and now :)

And I am always keeping my eye on some tasty and spicy updates for the addon, even when Gladiatus reminds me of a cemetery ,-D.
